File: CSharp/v1/Buyers/FinalizedDeals/ListFinalizedDeals.cs
/* Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using Google.Apis.AuthorizedBuyersMarketplace.v1;
using Google.Apis.AuthorizedBuyersMarketplace.v1.Data;
using Mono.Options;

using System;
using System.Collections.Generic;

namespace Google.Apis.AuthorizedBuyersMarketplace.Examples.v1.Buyers.FinalizedDeals
{
    /// <summary>
    /// Lists finalized deals for a given buyer and their clients.
    /// </summary>
    public class ListFinalizedDeals : ExampleBase
    {
        private AuthorizedBuyersMarketplaceService mkService;

        /// <summary>
        /// Constructor.
        /// </summary>
        public ListFinalizedDeals()
        {
            mkService = Utilities.GetAuthorizedBuyersMarketplaceService();
        }

        /// <summary>
        /// Returns a description about the code example.
        /// </summary>
        public override string Description
        {
            get => "This code example lists finalized deals for a given buyer and their clients.";
        }

        /// <summary>
        /// Parse specified arguments.
        /// </summary>
        protected override Dictionary<string, object> ParseArguments(List<string> exampleArgs) {
            string[] requiredOptions = new string[] {"account_id"};
            bool showHelp = false;

            string accountId = null;
            string filter = null;
            string orderBy = null;
            int? pageSize = null;

            string defaultFilter = "deal.dealType = PROGRAMMATIC_GUARANTEED";
            string defaultOrderBy = "deal.flightStartTime desc";

            OptionSet options = new OptionSet {
                "List finalized deals for the given buyer account and their clients.",
                {
                    "h|help",
                    "Show help message and exit.",
                    h => showHelp = h != null
                },
                {
                    "a|account_id=",
                    ("[Required] The resource ID of the buyers resource under which the " +
                     "finalized deals are being retrieved. This will be used to construct the " +
                     "parent used as a path parameter for the finalizedDeals.list request."),
                    a => accountId = a
                },
                {
                    "f|filter=",
                    ("Query string to filter finalized deals. By default, this example will " +
                     "fitler by deal type to retrieve programmatic guaranteed deals to " +
                     "demonstrate usage."),
                    f => filter =  f
                },
                {
                    "o|order_by=",
                    ("Query string used to sort the response of the list method. By default, " +
                     "this example will return deals in descending order of their flight start " +
                     "time to demonstrate usage. To learn more about the syntax for this " +
                     "parameter, see: " +
                     "https://cloud.google.com/apis/design/design_patterns#sorting_order"),
                    o => orderBy =  o
                },
                {
                    "p|page_size=",
                    ("The number of rows to return per page. The server may return fewer rows " +
                     "than specified."),
                    (int p) => pageSize =  p
                },
            };

            List<string> extras = options.Parse(exampleArgs);
            var parsedArgs = new Dictionary<string, object>();

            // Show help message.
            if (showHelp == true)
            {
                options.WriteOptionDescriptions(Console.Out);
                Environment.Exit(0);
            }
            // Set arguments.
            parsedArgs["account_id"] = accountId;
            parsedArgs["filter"] = filter ?? defaultFilter;
            parsedArgs["order_by"] = orderBy ?? defaultOrderBy;
            parsedArgs["pageSize"] = pageSize ?? Utilities.MAX_PAGE_SIZE;
            // Validate that options were set correctly.
            Utilities.ValidateOptions(options, parsedArgs, requiredOptions, extras);

            return parsedArgs;
        }

        /// <summary>
        /// Run the example.
        /// </summary>
        /// <param name="parsedArgs">Parsed arguments for the example.</param>
        protected override void Run(Dictionary<string, object> parsedArgs)
        {
            string accountId = (string) parsedArgs["account_id"];
            string parent = $"buyers/{accountId}";
            string pageToken = null;

            Console.WriteLine(@"Listing finalized deals for buyer account ""{0}""", parent);
            do
            {
                BuyersResource.FinalizedDealsResource.ListRequest request =
                    mkService.Buyers.FinalizedDeals.List(parent);
                request.Filter = (string) parsedArgs["filter"];
                request.OrderBy = (string) parsedArgs["order_by"];
                request.PageSize = (int) parsedArgs["pageSize"];
                request.PageToken = pageToken;

                ListFinalizedDealsResponse page = null;

                try
                {
                    page = request.Execute();
                }
                catch (Exception exception)
                {
                    throw new ApplicationException(
                        $"Marketplace API returned error response:\n{exception.Message}");
                }

                var finalizedDeals = page.FinalizedDeals;
                pageToken = page.NextPageToken;

                if (finalizedDeals == null)
                {
                    Console.WriteLine("No finalized deals found for buyer account.");
                }
                else
                {
                    foreach (FinalizedDeal finalizedDeal in finalizedDeals)
                    {
                        Utilities.PrintFinalizedDeal(finalizedDeal);
                    }
                }
            }
            while(pageToken != null);
        }
    }
}

File: ruby/v1/buyers/finalized_deals/list_finalized_deals.rb
#!/usr/bin/env ruby
# Encoding: utf-8
#
# Copyright:: Copyright 2022 Google LLC
#
# License:: Licensed under the Apache License, Version 2.0 (the "License");
#           you may not use this file except in compliance with the License.
#           You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#           Unless required by applicable law or agreed to in writing, software
#           distributed under the License is distributed on an "AS IS" BASIS,
#           WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#           implied.
#           See the License for the specific language governing permissions and
#           limitations under the License.
#
# Lists finalized deals for the given buyers and their clients.

require 'optparse'

require_relative '../../../util'


def list_finalized_deals(marketplace, options)
  name = "buyers/#{options[:account_id]}"

  page_token = nil

  puts "Listing finalized deals for buyer with name: \"#{name}\"."
  begin
    response = marketplace.list_buyer_finalized_deals(
        name,
        filter: options[:filter],
        order_by: options[:order_by],
        page_size: options[:page_size],
        page_token: page_token
    )

    page_token = response.next_page_token

    unless response.finalized_deals.nil?
      response.finalized_deals.each do |finalized_deal|
        FinalizedDealPrinter.print(finalized_deal)
      end
    else
      puts 'No finalized deals found.'
    end
  end until page_token.nil?
end


if __FILE__ == $0
  begin
    # Retrieve the service used to make API requests.
    service = get_service()
  rescue ArgumentError => e
    raise 'Unable to create service, with error message: #{e.message}'
  rescue Signet::AuthorizationError => e
    raise 'Unable to create service, was the KEY_FILE in util.rb set? Error message: #{e.message}'
  end

  # Set options and default values for fields used in this example.
  options = [
    Option.new(
      'account_id',
      'The resource ID of the buyers resource for which the finalized deals are being retrieved. This will be used '\
      'to construct the parent used as a path parameter for the finalizedDeals.list request.',
      type: Integer, short_alias: 'a', required: true, default_value: nil
    ),
    Option.new(
      'filter',
      'Query string to filter finalized deals. By default, this example will filter by deal type to retrieve '\
      'finalized deals including programmatic guaranteed deals to demonstrage usage.',
      type: Array, short_alias: 'f', required: false, default_value: 'deal.dealType = PROGRAMMATIC_GUARANTEED'
    ),
    Option.new(
      'order_by',
      'Query string used to sort the response of the list method. By default, this will return deals in descending '\
      'order of their flight start time to demonstrate usage. To learn more about the syntax for this parameter, '\
      'see: https://cloud.google.com/apis/design/design_patterns#sorting_order',
      type: Array, short_alias: 'f', required: false, default_value: 'deal.flightStartTime desc'
    ),
    Option.new(
      'page_size', 'The number of rows to return per page. The server may return fewer rows than specified.',
      type: Array, short_alias: 'p', required: false, default_value: MAX_PAGE_SIZE
    ),
  ]

  # Parse options.
  parser = Parser.new(options)
  opts = parser.parse(ARGV)

  begin
    list_finalized_deals(service, opts)
  rescue Google::Apis::ServerError => e
    raise "The following server error occured:\n#{e.message}"
  rescue Google::Apis::ClientError => e
    puts "Error: #{e}"
    raise "Invalid client request:\n#{e.message}"
  rescue Google::Apis::AuthorizationError => e
    raise "Authorization error occured:\n#{e.message}"
  end
end
